Add a guide on writing PR descriptions

frankot-inf opened the pull request:
Here's a Gist for easier reading: https://gist.github.com/frankot-inf/9dda1a98bba1e7c6fc5d0a35ca770eb4

There's also been a 'Reviewing a pull request' section in here up until now. However, as it's currently written, it doesn't seem like a fit.

> 1. When reviewing a pull request, make sure to check how the changes fit the current code base and architecture. Do not approve the pull request if the changes are not aligned with the rest of the codebase for no obvious reason.

In my opinion, this is almost equivalent to saying "when reviewing a pull request, review a pull request". 😄 

> 2. In addition to code inspection, check out the branch on your machine and test it on your mobile phone. Use common sense for this. It is not important to test the changes if a pull request changes only one file, but in case of some bigger features or fixes, you should test the implementation. The idea behind this is to catch some obvious issues (paddings, margins, user-specific bugs, etc.) before they reach our testers.

Hm, does anyone do this when reviewing a PR? I don't have that impression, especially when the reviewer is scheduled on a different project.

Not sure what to do with this. Should I keep it in?

ffloreani commented on Building quality apps/Pull requests.md:
```diff
@@ -1,19 +1,72 @@
+
+- [https://github.com/zplesac/android_connectionbuddy/pull/91](https://github.com/zplesac/android_connectionbuddy/pull/91)
+
+**HAK**:
+
+- [https://github.com/infinum/Android-Hak-Ispitivaci/pull/453](https://github.com/infinum/Android-Hak-Ispitivaci/pull/453)
+- [https://github.com/infinum/Android-Hak-Ispitivaci/pull/454](https://github.com/infinum/Android-Hak-Ispitivaci/pull/454)
+- [https://github.com/infinum/Android-Hak-Ispitivaci/pull/481](https://github.com/infinum/Android-Hak-Ispitivaci/pull/481)
+
+**GO**:
+
+- [https://github.com/infinum/Android-MyGO/pull/517](https://github.com/infinum/Android-MyGO/pull/517)
+- [https://github.com/infinum/Android-MyGO/pull/525](https://github.com/infinum/Android-MyGO/pull/525)
+- [https://github.com/infinum/Android-MyGO/pull/526](https://github.com/infinum/Android-MyGO/pull/526)
+- [https://github.com/infinum/Android-MyGO/pull/521](https://github.com/infinum/Android-MyGO/pull/521)
```
I like the idea of having this chapter in the public handbook, but if we're going public we should remove links to our private repos 🙂

frankot-inf replied:
Oh. I totally forgot that this was a public handbook.

ffloreani commented on Building quality apps/Pull requests.md:
```diff
@@ -1,19 +1,72 @@
+
+How much time should you invest in writing a PR description? Use your own reasoning on a case-by-case basis, as with everything. For very small PRs, it will take you a couple of minutes. For a small PR, 15 minutes should be sufficient. For a moderate PR, it's perfectly fine to take 30 minutes. For larger PRs, you may need an hour, or more. The return on investment is huge in either case. The more ...
+
+I believe that some examples can be useful. I'll use my own PRs, simply because I was trying to hold myself up to a standard. They are by no means perfect. Form your own opinion about them. Pick up what you like, and change what you don't. Try to review them, as a matter of fact. (Please don't leave actual comments.) See how it feels going through the files changed, or going through the commits.  ...
```
The personal note is kinda cool and I like the whole try-it-yourself concept, but we don't usually use the 1st person in this handbook. Maybe there's a way we could rewrite this to be a bit more 3rd person-ish? 😅

frankot-inf replied:
Yeah, I see. If we're going to have to remove the examples, this paragraph goes away anyway, though.

frankot-inf commented:
@BraneRi I can agree that in some of the linked PRs perhaps I was being overly detailed. I was still trying to find a balance and see what works and what doesn't.

However, in this specific case, I disagree.

New interactors have been added, and this needed to be mentioned. It clearly states, in one sentence, why they've been added, what their purpose is and what their relationship to the related existing interactors was.

Because the changes are unordered, a reviewer may stumble upon the usage of `SubmitResultInteractor` before its actual definition. The reviewer could perhaps remember that there exists a `DrivingResultInteractor` and wonder what's going on. They would figure it out, of course, but the point is that they shouldn't have to go back and forth to gather pieces of information. Well, the goal is to minimise that, at least.

BraneRi commented:
> @BraneRi I can agree that in some of the linked PRs perhaps I was being overly detailed. I was still trying to find a balance and see what works and what doesn't.
> 
> However, in this specific case, I disagree.
> 
> New interactors have been added, and this needed to be mentioned. It clearly states, in one sentence, why they've been added, what their purpose is and what their relationship to the related existing interactors was.
> 
> Because the changes are unordered, a reviewer may stumble upon the usage of `SubmitResultInteractor` before its actual definition. The reviewer could perhaps remember that there exists a `DrivingResultInteractor` and wonder what's going on. They would figure it out, of course, but the point is that they shouldn't have to go back and forth to gather pieces of information. Well, the goal is to minimise that, at least.

That makes sense. 

What I was trying to point out is danger of compensating code readability with PR descriptions. Because if reviewer is put into mental set that you well described in PR, the code will definitely make sense. But some new developer who joins the project hasn't seen this description and he/she will be in very different mental set.

This doesn't mean that we shouldn't write good PR descriptions, I just think that implementation specifics should be clear from implementation, not the description.

frankot-inf commented:
@BraneRi Of course. The reviewer should hold up their end - they should review.

Writing a description can also help the author figure out that the implementation is lacking in some respect, just by having a hard time writing about it.

itsJoKr commented:
Good job.

I'm looking at this your PR: https://github.com/infinum/Android-Hak-Ispitivaci
It's very well explained in detail, but then when I'm looking at the code I see barely two comment lines. And that's what is worrying me.

In my opinion, most of that PR you've nicely written should also be contained in the code documentation or some wiki page. The PR will be merged/closed and that's it. If someone comes to work on the project in a month or two, nobody is going back to read PRs. We want to figure out stuff from the source code.

It doesn't have to be perfect, but you get my point. This is much more valuable than writing complex PRs. And it can be read as a part of the PR. Here's one example from my project:

![image](https://user-images.githubusercontent.com/11093480/82569948-388be480-9b81-11ea-894b-26b4cdf0b9f5.png)

frankot-inf commented:
@itsJoKr You raise a good point. A good part of these PRs could definitely serve as a documentation as well. I'll try to do that from now on.

I want to affirm, though, that these should not be thought of as mutually exclusive. An important point in my guide is that the PR itself is a list of unordered changes by default. A list of random well-documented changes is still random.

itsJoKr commented:
Well said. Agreed!

frankot-inf commented:
@Aksi0m  Of course, always apply your own reasoning, I totally agree. I'll add more emphasis on this to the guide somewhere at the end.

**If** you are introducing a small amount of changes, **and if** your changes are well-documented in the code or are straight-up-obvious, **and if** your commit messages are small, focused, well-written and can stand for themselves, then a tiny description is probably more than enough.

If you have 100 files changed due to a rename, moving packages or similar, one sentence is probably enough. This actually follows directly from the algorithm, even. You'll introduce a 'section' with a single sentence explaining the rename, e.g. pointing out a typo. All other changes fall under this same goal and are related, nothing needs to be added.

Asking yourself 'Would I have a hard time reviewing this PR if I was the new guy on the project?' is a good but very undefined and subjective question. Better would be to give the algorithm a pass, and see what comes out, in my opinion. If it's really one of those cases, then you'll be done with it in 5 minutes or less.

I think that, in an average case, a description is needed, rather than not needed.

frankot-inf commented:
The examples I've provided may be misleading. The biggest problem with them is their size. They have such a description because they are big, and a big description automatically follows. Likewise, as Josip mentioned, chunks of the descriptions should have probably been included in docstrings and GitHub wiki pages, which should also be referenced from the code. This way a lot of the description would just be solving the problem of the PR being an unordered list of changes, and would simply reference all the existing information somewhere else.

frankot-inf commented on Building quality apps/Pull requests.md:
```diff
@@ -1,19 +1,72 @@
+
+The title should follow the same format as a commit message heading. That is, use the imperative mood, as if giving a command, and keep it short. You can leave the branch name, if you choose to, but it takes 10 seconds to think of something better. The title is far less important than the description, though.
+
+How much time should you invest in writing a PR description? Use your own reasoning on a case-by-case basis, as with everything. For very small PRs, it will take you a couple of minutes. For a small PR, 15 minutes should be sufficient. For a moderate PR, it's perfectly fine to take 30 minutes. For larger PRs, you may need an hour, or more. The return on investment is huge in either case. The more ...
```
I mainly wrote that to erase the potential fear of taking too much time. If you need X amount of time to do it, then you need X amount of time. I might reword this.

If most of the PR you wrote can serve as the documentation, then it doesn't matter whether you spent the time writing that up in the PR or before that, provided of course that you go ahead and include that information back in the code or Wiki. (Which I didn't do in my PRs until now.)

And yes, *if* you provide all that information in code or wiki, *then* writing the PR description will take a lot less time.

I am yet to see a PR with all the useful information in the code or wiki and a satisfactory description, though.

frankot-inf commented:
> Also, I guess when you are working full time on some project with another developer and you both review each other PRs I don't think you need a lot of PR description if any. Since the reviewer works on the same project he understands the background of that PR and the wider picture, for example when I worked on NutriU, task link in the description was enough in 90% of cases just to give you information for what that PR is since you are in a loop and tasks were defined and written very good and detailed.

Yes, as I've said several times by now, use your own reasoning. If you know the reviewer understands the background, then you don't need to provide that.

Link is not enough by default. From the guide:

> A task specifies what needs to be done, but the reviewer needs to review the implementation of it, and the description of the implementation needs to exist.

And literally the entire 'Contextualise and group changes together' section was there to explain further. 😄 

Again, yes, if the PR contains 2 files changed, then you don't need anything. I'm talking about the general, average case.

frankot-inf commented on Building quality apps/Pull requests.md:
```diff
@@ -1,19 +1,72 @@
 
-1. When reviewing a pull request, make sure to check how the changes fit the current code base and architecture. Do not approve the pull request if the changes are not aligned with the rest of the codebase for no obvious reason.
-2. In addition to code inspection, check out the branch on your machine and test it on your mobile phone. Use common sense for this. It is not important to test the changes if a pull request changes only one file, but in case of some bigger features or fixes, you should test the implementation. The idea behind this is to catch some obvious issues (paddings, margins, user-specific bugs, etc.) befo ...
+Links to tasks can be helpful, but they are very far from enough. Even if the task is clearly defined and easy to understand, which is not always the case, tasks are more useful to the developer than the reviewer. A task specifies what needs to be done, but the reviewer needs to review the *implementation* of it, and the description of the *implementation* needs to exist.
```
Yes, I wanted to mention links to design, but forgot to. Good catch. 👍

ffloreani commented:
I've converted this to a draft PR as it seems to be still of a work in progress (and that's perfectly OK). Once you complete all your changes, mark it as _Ready for review_ and we'll have another look at it 🙂

frankot-inf commented:
- Removed links to our private repositories and the accompanying paragraph.
- Removed the paragraph about the amount of time to spend.
- Added more emphasis on approaching PRs on a case-by-case basis.
- Added suggestions to use docstrings and GitHub Wiki pages.
- Added a suggestion to provide a link or screenshots of the UI design, and potentially its implementation.
